src/resolvers/mutation.ts

```diff
@@ -300,6 +300,50 @@ export async function createTags(
   return tags;
 }
 
+/**
+ * Creates tags, savedItem and tag association for the given list of inputs.
+ * Note: this mutation does not validate if the savedItem exist already.
+ * @param root
+ * @param args list of savedItemTagsInput.
+ * @param context
+ */
+export async function createSavedItemTags(
+  root,
+  args: { input: SavedItemTagsInput[] },
+  context: IContext
+): Promise<SavedItem[]> {
+  const savedItemService = new SavedItemDataService(context);
+  const tagDataService = new TagDataService(context, savedItemService);
+
+  const tagCreateInput: TagCreateInput[] = [];
+  for (const savedItemTagsInput of args.input) {
+    for (const tag of savedItemTagsInput.tags) {
+      tagCreateInput.push({
+        name: tag,
+        savedItemId: savedItemTagsInput.savedItemId,
+      });
+    }
+  }
+
+  await tagDataService.insertTags(tagCreateInput);
+  const savedItemIds: string[] = args.input.map((i) => i.savedItemId);
+  const savedItems = await savedItemService.batchGetSavedItemsByGivenIds(
+    savedItemIds
+  );
+
+  const savedItemTagsMap = getSavedItemTagsMap(args.input);
+
+  for (const savedItem of savedItems) {
+    context.emitItemEvent(
+      EventType.ADD_TAGS,
+      savedItem,
+      savedItemTagsMap[savedItem.id]
+    );
+  }
```

Comment:
I think this can be simplified. `savedItemIds` isn't deduplicated, so there's not really a reason to call `getSavedItemTagsMap`. You can just use `SavedItemTagsInput` directly to emit the events.

Reply:
the goal for calling `getSavedItemTagsMap` is to remove any duplicate savedItems and duplicate tags within savedItems, so we don't emit these duplicates in the snowplow events.

i will refactor the `getSavedItemTagsMap` function to see how we can remove duplicates in-memory

Reply:
If we want to deduplicate and aggregate the response, we should probably do it before creating the tags and requesting back the saved items

Reply:
oh deduplication is done inside the `insertTags` function, this one does the same thing before sending the snowplow event

Reply:
I'm wrong! no bug